homeassistant/components/nibe_heatpump/__init__.py

```diff
@@ -0,0 +1,242 @@
+            model=heatpump.model.name,
+            name=heatpump.model.name,
+        )
+
+        await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    except Exception:
+        await connection.stop()
+        raise
+
+    # Trigger a refresh again now that all platforms have registered
+    hass.async_add_job(coordinator.async_refresh)
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        coordinator: Coordinator = hass.data[DOMAIN].pop(entry.entry_id)
+        await coordinator.connection.stop()
+
+        # sleep a bit to make sure sockets are closed
+        await asyncio.sleep(5)
+
+    return unload_ok
+
+
+class Coordinator(DataUpdateCoordinator[dict[int, Coil]]):
+    """Update coordinator for nibe heat pumps."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        heatpump: HeatPump,
+        connection: Connection,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name="Nibe Heat Pump", update_interval=timedelta(seconds=60)
+        )
+
+        self.connection = connection
+        self.heatpump = heatpump
+
+    @property
+    def coils(self) -> dict[str, Coil]:
+        """Return the full coil database."""
+        return self.heatpump._address_to_coil  # pylint: disable=protected-access
+
+    @property
+    def unique_id(self) -> str:
+        """Return unique id for this coordinator."""
+        assert self.config_entry
+        return self.config_entry.unique_id or self.config_entry.entry_id
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information for the main device."""
+        assert self.config_entry
+        return DeviceInfo(identifiers={(DOMAIN, self.unique_id)})
+
+    def get_coil_data(self, coil: Coil | None) -> Coil | None:
+        """Return a coil with data and check for validity."""
+        if not self.data or not coil:
+            return None
+
+        coil = self.data.get(coil.address)
+        if coil and coil.value != -3276.8:
+            return coil
+        return None
+
+    def get_coil_value(self, coil: Coil | None) -> int | str | float | None:
+        """Return a coil with data and check for validity."""
+        if coil := self.get_coil_data(coil):
+            return coil.value
+        return None
+
+    def get_coil_float(self, coil: Coil | None) -> float | None:
+        """Return a coil with float and check for validity."""
+        if value := self.get_coil_value(coil):
+            return float(value)
+        return None
+
+    async def async_write_coil(
+        self, coil: Coil | None, value: int | float | str
+    ) -> None:
+        """Write coil and update state."""
+        if not coil:
+            raise Exception("No coil available")
```

Comment:
I would recommend creating a more explicit exception for this; `NoCoilAvailableError`?

Reply:
I switched to HomeAssistantError() it's something that is possibly shown in UI, not expected to be caught by anything other than some generic handler.